### ISSUE TITLE
preInit (and maybe onRuntimeInitialized?) should support Promises/async

### ISSUE THREAD

**Issue by kg** [emscripten-core/emscripten#14520]
Because of how async works in JS (your function just starts returning a Promise) it's easy for someone to make a preInit or onRuntimeInitialized handler async and not notice that the behavior is subtly wrong, introducing a race condition. Handling this by hand using runtime dependencies is a little tricky because of how promise fulfillment works (next event loop turn) vs how runtime dependencies work (as soon as they're all gone, the callback is invoked synchronously), so you can accidentally introduce a new race when doing it. If preInit (and potentially onRuntimeInitialized) were updated to notice if the handler returned a thenable it would be possible for all this to just work and it would reduce the risk of people adding races by misusing runtime dependencies or forgetting about them entirely.

I think implementing this would just mean changing the loop at the bottom of generated JS that runs preInit functions so that it notices thenables and chains off .then if it encounters one, and then do the final run invoke in the last callback.

**Comment by kripken** [comment]
I agree those options are not optimal, yeah. Another option here is to use Modularize, as mentioned near the end of this FAQ entry, as the Modularize factory function returns a proper Promise now,

https://emscripten.org/docs/getting_started/FAQ.html#how-can-i-tell-when-the-page-is-fully-loaded-and-it-is-safe-to-call-compiled-functions

Is that good enough in general, do you think?

**Comment by sbc100** [comment]
This seems like a good idea to me.   One of the issues makes this kind of stuff complex (IIUC) is that we have  single codebase the tries to support both sync and async paths though it (due to the WASM_ASYNC_COMPILATION=0 options which I wish we could remove but seems here to stay).

**Comment by sbc100** [comment]
Also hi Katelyn!

**Comment by kg** [comment]
> 
> 
> This seems like a good idea to me. One of the issues makes this kind of stuff complex (IIUC) is that we have single codebase the tries to support both sync and async paths though it (due to the WASM_ASYNC_COMPILATION=0 options which I wish we could remove but seems here to stay).

Yeah, it's very tricky. I think if this is done correctly it won't have any visible impact on existing code, and you'll only 'pay' for it if one of the callbacks returns a thenable object, at which point the startup code just chains off it. It might have subtle side effects I'm not thinking of, but at least the user is opting into it by making their callbacks return a promise. Maybe if ```WASM_ASYNC_COMPILATION=0```, instead of chaining off the .then method it throws an Error to alert you that you've done something that doesn't make sense.

**Comment by kripken** [comment]
I'm in favor of this in general, but I'm not sure if we can make it not break existing users that depend on the current behavior. If we can, though, then it sounds very good. Definitely worth looking into.

**Comment by sbc100** [comment]
Isn't it just a question of adding additional code to handle the return value of the user callbacks?   If the users doesn't return a promise they shouldn't see any change at all.   The only downside is that extra code needed to handle this use case.

**Comment by kg** [comment]
I wouldn't be surprised if some existing code accidentally returns a promise and is relying on the way it works right now.
